### PR TITLE
Added note about immutability.

### LINKED ISSUE #12: Communicate immutability
We should communicate that we use immutable objects and as such two objects of the same kind might have the same reference.

### PR DESCRIPTION
**Pull request by Nyholm**
Is this a better place to warn about immutability?

This will fix #12

### CODE REVIEW

File: proposed/http-client/http-client.md [hunk 1]
```diff
@@ -46,6 +46,10 @@ interface HttpClient
      *
      * If a request is sent without any prior configuration, a HTTP client MUST NOT follow redirects.
      *
+     * The client MAY do motifications to the Request before sending it. Because of PSR-7 objects are
+     * immutable one cannot assume that the object passed to HttpClient::sendRequest will have the same
+     * reference as the Request object returned by an exception.
```
Comment (sagikazarmark) [inline proposed/http-client/http-client.md]: I would not mention exception specifically as the user might access the messages  from other sources. I would rather say something more general and mention exceptions as an example.

**Review by sagikazarmark — CHANGES_REQUESTED** [review]

File: proposed/http-client/http-client.md [hunk 2]
```diff
@@ -46,6 +46,11 @@ interface HttpClient
      *
      * If a request is sent without any prior configuration, a HTTP client MUST NOT follow redirects.
      *
+     * The client MAY do motifications to the Request before sending it. Because of PSR-7 objects are
```
Comment (dbu) [inline proposed/http-client/http-client.md]: /Because of PSR-7 objects are/Because PSR-7 objects are/

File: proposed/http-client/http-client.md [hunk 3]
```diff
@@ -46,6 +46,11 @@ interface HttpClient
...
+     * immutable one cannot assume that the object passed to HttpClient::sendRequest will have the same
```
Comment (dbu) [inline proposed/http-client/http-client.md]: immutable, one

File: proposed/http-client/http-client.md [hunk 4]
```diff
@@ -46,6 +46,11 @@ interface HttpClient
...
+     * reference as the Request object that is actually sent. For example the Request object that is
```
Comment (dbu) [inline proposed/http-client/http-client.md]: is reference the right word? could we just say "will be the same object"?

File: proposed/http-client/http-client.md [hunk 5]
```diff
@@ -46,6 +46,11 @@ interface HttpClient
      *
      * If a request is sent without any prior configuration, a HTTP client MUST NOT follow redirects.
      *
+     * The client MAY do motifications to the Request before sending it. Because PSR-7 objects are
+     * immutable, one cannot assume that the object passed to HttpClient::sendRequest will be the same
+     * object that is actually sent. For example the Request object that is returned by an exception MAY
+     * have a different reference then the object passed here.
```
Comment (dbu) [inline proposed/http-client/http-client.md]: here you still say reference. maybe "the Request object that is returned ... MAY be a different object than the one passed to sendRequest, so `===` comparison is not possible.

and its "different than", not "different then"

**Comment by Nyholm** [comment]
Thank you for the reviews. This PR is updated.

**Review by dbu — APPROVED** [review]
Looks good to me

File: proposed/http-client/http-client.md [hunk 6]
```diff
@@ -46,6 +46,11 @@ interface HttpClient
      *
      * If a request is sent without any prior configuration, a HTTP client MUST NOT follow redirects.
      *
+     * The client MAY do motifications to the Request before sending it. Because PSR-7 objects are
+     * immutable, one cannot assume that the object passed to HttpClient::sendRequest will be the same
+     * object that is actually sent. For example the Request object that is returned by an exception MAY
+     * be a different object than the one passed to sendRequest, so === comparison is not possible.
```
Comment (sagikazarmark) [inline proposed/http-client/http-client.md]: maybe say comparison by reference?

File: proposed/http-client/http-client.md [hunk 7]
```diff
@@ -46,6 +46,11 @@ interface HttpClient
...
+     *
```
Comment (sagikazarmark) [inline proposed/http-client/http-client.md]: Maybe place a link to the relevant part of PSR-7?

https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-7-http-message-meta.md#why-value-objects

Although not necessarily immutability is the main point there, but explains the WHY.

**Review by sagikazarmark — CHANGES_REQUESTED** [review]

File: proposed/http-client/http-client.md (hunk 6 above)
Comment (Nyholm) [reply proposed/http-client/http-client.md]: @dbu whats to move away from the word "reference" (https://github.com/php-http/fig-standards/pull/16#discussion_r80212953). I have no issues using that word as long as it makes sense and is easy to understand.

Comment (sagikazarmark) [reply proposed/http-client/http-client.md]: Hm, @dbu what's the downside of using the word "reference"? For me === comparison sounds like phrasing the same for people with less knowledge.

Comment (dbu) [reply proposed/http-client/http-client.md]: lets use reference. depending on how beginner-friendly we want to make it, we could have the === in parenthesis. but go with what you prefer @Nyholm. the important thing is that we mention, and people that are unsure what reference is probably won't get what we try to tell them anyways and have to figure it out the hard way. and anyways all existing PSR-7 implementations will likely already have that behaviour so we just better document the only possible way things can be with PSR-7

Comment (Nyholm) [reply proposed/http-client/http-client.md]: Thank you

**Comment by Nyholm** [comment]
Thank you for your reviews

**Pull request merged**